lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs

```diff
@@ -0,0 +1,692 @@
+                    Err(invalid_data_error(&format!(
+                        "got unsuccessful ServerDeviceAnnounceResponse result code NTSTATUS({})",
+                        &req.result_code
+                    )))
+                } else {
+                    Ok(())
+                }
+            }
+            PacketId::PAKID_CORE_DEVICE_IOREQUEST => {
+                let req = DeviceIoRequest::decode(&mut payload)?;
+                debug!("got {:?}", req);
+
+                if let MajorFunction::IRP_MJ_DEVICE_CONTROL = req.major_function {
+                    let ioctl = DeviceControlRequest::decode(req, &mut payload)?;
+                    debug!("got {:?}", ioctl);
+
+                    let (code, res) = self.scard.ioctl(ioctl.io_control_code, payload)?;
+                    if code == SPECIAL_NO_RESPONSE {
+                        return Ok(());
+                    }
+                    let resp = encode_message(
+                        PacketId::PAKID_CORE_DEVICE_IOCOMPLETION,
+                        &mut DeviceControlResponse::new(&ioctl, code, res).encode()?,
+                    )?;
+                    debug!("sending device IO response");
+                    Ok(mcs.write(&"rdpdr".to_string(), resp)?)
+                } else {
+                    Err(invalid_data_error(&format!(
+                        "got unsupported major_function in DeviceIoRequest: {:?}",
+                        &req.major_function
+                    )))
+                }
+            }
+            _ => {
+                // TODO(awly): return an error here once the entire protocol is implemented?
+                error!(
+                    "RDPDR packets {:?} are not implemented yet, ignoring",
+                    header.packet_id
+                );
+                Ok(())
+            }
+        }
+    }
+}
+
+fn encode_message(packet_id: PacketId, payload: &mut Vec<u8>) -> RdpResult<Vec<u8>> {
+    let mut inner = Header::new(Component::RDPDR_CTYP_CORE, packet_id).encode()?;
+    inner.append(payload);
+    let mut outer = ChannelPDUHeader::new(inner.length() as u32).encode()?;
+    outer.append(&mut inner);
+    Ok(outer)
+}
+
+// ChannelPDUHeader flags.
+const CHANNEL_FLAG_FIRST: u32 = 0x00000001;
+const CHANNEL_FLAG_LAST: u32 = 0x00000002;
+
+#[derive(Debug)]
+struct ChannelPDUHeader {
+    length: u32,
+    flags: u32,
+}
+
+impl ChannelPDUHeader {
+    fn new(length: u32) -> Self {
+        Self {
+            length,
+            flags: CHANNEL_FLAG_FIRST | CHANNEL_FLAG_LAST,
+        }
+    }
+    fn decode(payload: &mut Payload) -> RdpResult<Self> {
+        Ok(Self {
+            length: payload.read_u32::<LittleEndian>()?,
+            flags: payload.read_u32::<LittleEndian>()?,
+        })
+    }
+    fn encode(&self) -> RdpResult<Vec<u8>> {
+        let mut w = vec![];
+        w.write_u32::<LittleEndian>(self.length)?;
+        w.write_u32::<LittleEndian>(self.flags)?;
+        Ok(w)
+    }
```

Comment:
Only use a Vec instead of a statically allocated `std::io::Cursor<&mut [u8]>`or similar here if you must. If not, preallocate the Vec using `Vec::with_capacity` with an appropriate estimated capacity. In general, prefer using cursors with arrays when you can predict the total size you will need. If you cannot know, use a Vec with a minimum expected usage using `Vec::with_capacity`.